In [1]:
import pandas as pd
import numpy as np

In [4]:
df = pd.read_csv(r"C:\Users\Junayed\nyc-311-resolution-time-analysis\data\clean\311_jan_week1_clean.csv")

In [5]:
df.shape

(85122, 11)

### Analyze Full Frequency Distribution of Resolution Turnaround

Evaluate the value count distribution of `Resolution_Days` across all 389 unique duration lengths:

* **High Front-End Concentration:** Over $65\%$ of cases (**55,687**) are resolved on the same day (`0` days), and over $79\%$ are completed within 24–48 hours (`0`–`1` days: **67,888** tickets).
* **Rapid Drop-Off:** Resolution counts decay sharply from Day 0 down through Day 4 (**917** tickets), illustrating standard operational throughput for high-volume routine complaints.
* **Extreme Tail Dispersion:** The existence of 389 distinct day counts with single-ticket outliers stretching up to several hundred days (e.g., 222, 285, 336, 503, 582) visually confirms a heavily right-skewed distribution driven by delayed administrative closures or complex field investigations.

In [13]:
df["Resolution_Days"].value_counts()

Resolution_Days
0      55687
1      12201
2       5024
3       2261
4        917
       ...  
222        1
582        1
336        1
285        1
503        1
Name: count, Length: 389, dtype: int64

### Inspect Top 10 Most Frequent Resolution Durations

Examine the highest-frequency turnaround times to understand operational velocity across typical service requests:

* **Dominant Velocity:** Turnaround times of $0$ to $2$ days comprise the vast majority of the top tiers, led heavily by same-day closures (**55,687** tickets) and 1-day completions (**12,201** tickets).
* **Decay Trajectory:** Volume decreases monotonically from Day 0 through Day 7, reflecting standard resolution curves where simpler operational tasks are cleared early in the ticket lifecycle.
* **Slight Non-Linearity (Days 8–9):** Day 9 (**369** tickets) marginally outpaces Day 8 (**317** tickets), likely reflecting post-weekend administrative batch closures or scheduled review cycles.

In [12]:
df["Resolution_Days"].value_counts().head(10)

Resolution_Days
0    55687
1    12201
2     5024
3     2261
4      917
5      594
6      567
7      509
9      369
8      317
Name: count, dtype: int64

In [ ]:
df["Resolution_Days"].value_counts().sort_index(ascending=False).head(10)

Resolution_Days
622    1
621    2
620    1
619    1
600    1
592    1
591    1
582    1
581    1
576    1
Name: count, dtype: int64

### Analyze Complaint Type Frequency Distribution

Evaluate the overall volume and diversity of service request categories across all 133 unique complaint classifications:

* **Primary Drivers:** The top three categories alone account for the vast majority of intake:
  * **Noise - Residential:** Dominates the dataset with **30,224** records ($>35\%$ of total complaints).
  * **Heat/Hot Water:** Represents the second largest share with **14,711** records, typical for seasonal winter demand in early January.
  * **Illegal Parking:** Logs **9,745** records, making quality-of-life and curb-space enforcement top municipal priorities.
* **Secondary Volume:** Issues like `Blocked Driveway` (**3,297**) and `Street Condition` (**1,310**) represent moderate recurring city service needs.
* **Long Tail Distribution:** The remaining categories taper off sharply into single-incident specialized complaints (e.g., `Scaffold Safety`, `Pet Shop`, `Building Condition`), illustrating broad operational scope across NYC agencies.

In [7]:
df["Complaint_Type"].value_counts()

Complaint_Type
Noise - Residential               30224
Heat/Hot Water                    14711
Illegal Parking                    9745
Blocked Driveway                   3297
Street Condition                   1310
                                  ...  
Posting Advertisement                 1
Pet Shop                              1
Building Condition                    1
Scaffold Safety                       1
Institution Disposal Complaint        1
Name: count, Length: 133, dtype: int64

### Inspect Top 10 Most Frequent Complaint Types

Examine the ten highest-volume service request categories to identify the primary municipal drivers across New York City:

* **Dominant Concerns:** Quality-of-life and residential comfort issues dominate citywide intake:
  * **Residential Living:** `Noise - Residential` (**30,224**) and `Heat/Hot Water` (**14,711**) alone account for over $52\%$ of all closed requests.
  * **Curb Management & Traffic:** `Illegal Parking` (**9,745**), `Blocked Driveway` (**3,297**), and `Abandoned Vehicle` (**1,195**) represent major street enforcement workloads.
* **Environmental & Infrastructure:** `Street Condition` (**1,310**), `Unsanitary Condition` (**1,273**), and `Water System` (**954**) reflect essential neighborhood infrastructure and sanitation maintenance.
* **Secondary Noise Channels:** Public and commercial noise complaints (`Noise - Commercial`: **1,058**; `Noise - Street/Sidewalk`: **1,054**) add further burden to municipal sound monitoring beyond residential properties.

In [14]:
df["Complaint_Type"].value_counts().head(10)

Complaint_Type
Noise - Residential        30224
Heat/Hot Water             14711
Illegal Parking             9745
Blocked Driveway            3297
Street Condition            1310
Unsanitary Condition        1273
Abandoned Vehicle           1195
Noise - Commercial          1058
Noise - Street/Sidewalk     1054
Water System                 954
Name: count, dtype: int64

### Final Geographic Distribution by Borough (Post-Cleaning)

Evaluate the final distribution of closed service requests across all five NYC boroughs following outlier removal, missing-value filtering, and deduplication:

* **Cohort Integrity:** The post-cleaning volume confirms **85,122** verified complaints fully attributed across the five boroughs with zero unspecified entries.
* **Volume Distribution:**
  * **Bronx:** Remains the primary source of service requests at **37,641** cases ($44.2\%$ of total volume).
  * **Brooklyn:** Represents the second-highest concentration with **18,859** complaints ($22.2\%$).
  * **Queens & Manhattan:** Account for **14,306** ($16.8\%$) and **12,243** ($14.4\%$) records, respectively.
  * **Staten Island:** Accounts for the smallest operational volume at **2,073** requests ($2.4\%$).
* **Consistency Check:** Proportional representation across boroughs remains stable compared to the initial raw intake, confirming that data hygiene filters did not introduce regional sampling bias.

In [8]:
df["Borough"].value_counts()

Borough
Bronx            37641
Brooklyn         18859
Queens           14306
Manhattan        12243
Staten Island     2073
Name: count, dtype: int64

### Analyze Service Request Volume by Responding Agency

Evaluate operational workload distribution across municipal departments handling closed 311 tickets:

* **Primary Responders:** Two agencies absorb over $80\%$ of total service request intake:
  * **NYPD (Police Department):** Leads citywide volume with **49,047** tickets ($57.6\%$), handling quality-of-life enforcement including residential/commercial noise, blocked driveways, and illegal parking.
  * **HPD (Housing Preservation & Development):** Manages **19,787** requests ($23.2\%$), predominantly addressing residential heating, hot water, and unsanitary building conditions.
* **Secondary Responders:**
  * **DSNY (Sanitation):** **4,674** tickets covering illegal dumping and collection concerns.
  * **DOT (Transportation):** **3,472** tickets addressing street, sidewalk, and traffic condition repairs.
  * **DEP (Environmental Protection):** **2,766** complaints for water system, sewer, and environmental noise issues.
* **Specialized Long Tail:** Specialized departments—including **DOB** (Buildings: 1,558), **DOHMH** (Health: 1,237), **DPR** (Parks: 898), **TLC** (Taxi & Limousine: 526), **EDC**, **DHS**, **DCWP**, **DOE**, and **OTI** (6 tickets)—manage low-volume, jurisdiction-specific cases.

In [9]:
df["Agency"].value_counts()

Agency
NYPD     49047
HPD      19787
DSNY      4674
DOT       3472
DEP       2766
DOB       1558
DOHMH     1237
DPR        898
TLC        526
EDC        452
DHS        355
DCWP       291
DOE         53
OTI          6
Name: count, dtype: int64